### PR TITLE
[Feature] Add cache to disk

### LINKED ISSUE #3092: [Feature]: Cache to disk
### The Feature

Currently, caching only supports in-memory cache, redis cache and s3 cache.

This feature would enable the library to store generated completions on disk and keep a persistent result of the generations after the interruption of my program.
A tool like [diskcache](https://grantjenks.com/docs/diskcache/) could be used.

### Motivation, pitch

I often run large batch completions. If an error occurs during the generation or I have to interrupt the program, I lose the results because they are not saved.

Disk Caching would help restart the batch completion from where it was interrupted.

### Twitter / LinkedIn details

@antonio_loison

### PR DESCRIPTION
**Pull request by antonioloison**
resolves #3092. 

Added disk cache using the `diskcache` module.

Please check the issue for more context.

### CODE REVIEW

File: litellm/tests/test_caching.py [hunk 1]
```diff
@@ -875,6 +875,66 @@ async def test_redis_cache_acompletion_stream_bedrock():
...
+        1, 100000
+    )  # add a random number to ensure it's always adding / reading from cache
+    messages = [
+        {"role": "user", "content": f"write a one sentence poem about: {random_number}"}
+    ]
+    litellm.cache = Cache(
+        type="disk",
+    )
+    print("test2 for Redis Caching - non streaming")
+    response1 = completion(
```
Comment (ishaan-jaff) [inline litellm/tests/test_caching.py:893]: can you use `mock_response` here so it does not make the LLM API call 

if you look up `mock_response` you'll see how it;s used through the testing pipeline

File: litellm/caching.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 from litellm._logging import verbose_logger
 from litellm.types.services import ServiceLoggerPayload, ServiceTypes
 import traceback
+import diskcache as dc
```
Comment (ishaan-jaff) [inline litellm/caching.py]: can this not be imported here, instead please import it inside the `DiskCache` class

Comment (ishaan-jaff) [reply litellm/caching.py]: We don't want this impacting users on litellm proxy server

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]

File: litellm/caching.py (hunk 2 above)
Comment (ishaan-jaff) [reply litellm/caching.py]: since this is an optional dependency it should only be imported in the class

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
left review, can you please add this to docs too. Thanks

File: litellm/tests/test_caching.py (hunk 1 above)
Comment (antonioloison) [reply litellm/tests/test_caching.py:893]: Thank you for the review
I just corrected it, you can resolve the conversation if you think it is correct, please

File: litellm/caching.py (hunk 2 above)
Comment (antonioloison) [reply litellm/caching.py]: I understand, I moved it to the class.
I removed the diskcache dependency from the `pyproject.toml` file and added the install command in the docs in the same way as the `s3-cache`.

**Comment by antonioloison** [comment]
> left review, can you please add this to docs too. Thanks

I added the doc for the disk cache and corrected the arguments of the `add_cache`, `update_cache`... functions.
I also changed the name of the file `redis_cache.md` to `all_caches.md` as it no longer only concerns the redis cache.

**Review by ishaan-jaff — APPROVED** [review]

**Pull request merged**